### PR TITLE
Avoid unsoundness in the use of the Pin API

### PR DESCRIPTION
**Pull request by mystor**

### CODE REVIEW

**Review by yaahc — APPROVED** [review]
looks good to me, I wanna get an approval from kenny too before I merge

cc @kennytm

File: fakecore/src/any.rs [hunk 1]
```diff
@@ -1,88 +1,95 @@
...
     }
 
     /// Get the `TypeId` of the requested type.
     pub fn type_id(&self) -> TypeId {
-        self.0.type_id
+        self.type_id
     }
 
     /// Returns `true` if the requested type is the same as `T`
     pub fn is<T: ?Sized + 'static>(&self) -> bool {
```
Comment (yaahc) [inline fakecore/src/any.rs:54]: I kinda wanna rename `is` to `is_requesting`

File: fakecore/src/any.rs [hunk 2]
```diff
@@ -1,88 +1,95 @@
...
     /// Calls the provided closure with a request for the the type `T`, returning
     /// `Some(&T)` if the request was fulfilled, and `None` otherwise.
-    ///
-    /// The `ObjectProviderExt` trait provides helper methods specifically for
-    /// types implementing `ObjectProvider`.
-    pub fn with<T: ?Sized + 'static, F>(f: F) -> Option<&'out T>
+    pub fn with<T: ?Sized + 'static, F>(f: F) -> Option<&'a T>
     where
-        F: FnOnce(Pin<&mut Self>),
+        F: FnOnce(Pin<&mut Request<'a>>),
```
Comment (kennytm) [inline fakecore/src/any.rs]: i think we could keep using `Self` here 🙂

**Review by kennytm — APPROVED** [review]
LGTM

File: fakecore/src/any.rs (hunk 2 above)
Comment (mystor) [inline fakecore/src/any.rs]:
```suggestion
        F: FnOnce(Pin<&mut Self>),
```

**Pull request merged**